x-pack/plugins/apm/server/lib/service_map/get_service_map_from_trace_ids.ts

```diff
@@ -35,7 +36,7 @@ export function getConnections(
             SERVICE_NAME in node &&
             (node as ServiceConnectionNode)[SERVICE_NAME] === serviceName;
         }
-        if (environment) {
+        if (environment && environment !== ENVIRONMENT_NOT_DEFINED.value) {
```

Comment:
Is this correct? Looks like we'll no filter when the user selects `not defined`, meaning `not defined` and `all` are treated the same.
What is the value of `not defined` when we get it from ES? Elsewhere we use the `missing` param and set it to `NOT_DEFINED`

Reply:
@sqren thanks for you question, I don't have that much knowledge on this code, but if I don't add this condition `&& environment !== ENVIRONMENT_NOT_DEFINED.value` the service maps shows like:
<img width="378" alt="Screenshot 2020-09-16 at 13 11 03" src="https://user-images.githubusercontent.com/55978943/93330063-7b538580-f81e-11ea-9222-e7439438804d.png">
and with the condition, it shows all connections:
<img width="150" alt="Screenshot 2020-09-16 at 13 12 30" src="https://user-images.githubusercontent.com/55978943/93330094-8ad2ce80-f81e-11ea-8921-db30afadce4a.png">

Reply:
@dgieselaar @smith @ogupte do you guys know if that's correct to add this condition here?

Reply:
Okay, let's leave this for now

Reply:
`ENVIRONMENT_NOT_DEFINED` is not a value that will be set on the documents from the scripted_metric aggregation, because it doesn't exist as such in the documents itself. You'd need something like:

```ts

if (environment === ENVIRONMENT_NOT_DEFINED.value) {
	matches = !('service.environment' in node);
} else {
	matches = 'service.environment' in node && node['service.environment'].value == environment;
}
```

(I haven't tried this out).

Reply:
@dgieselaar for the data we have, what you suggested will return the same result as the current implementation, that is because when `NOT_DEFINED` is selected the `node` variable contains the `service.environment` but it is set to `null`. Do you think is worth changing the current implementation to what you suggested even though?

Reply:
AFAICT, I think @sqren is right in the sense that this function should _always_ filter, unless `environment` is `ENVIRONMENT_ALL`. If `environment` is `ENVIRONMENT_NOT_DEFINED`, only paths that have at least one node where `service.environment` is `null` should be returned. Now, we are returning all paths when `service.environment` is `ENVIRONMENT_NOT_DEFINED`, which seems like a bug to me.

Reply:
So we should keep the old implementation: 
```js
 if (environment) {
```
instead of:
```js
if (environment && environment !== ENVIRONMENT_NOT_DEFINED.value) {
```

Reply:
No, because ENVIRONMENT_NOT_DEFINED doesn't exist in ES documents, so
you'll run into false negatives.

Reply:
Try this (untested)

```ts
      return path.some((node) => {
        // no match if service name is given and doesn't match
        if (serviceName && node[SERVICE_NAME] !== serviceName) {
          return false;
        }

        if (environment) {
          // no match if environment is "not defined" and the node has an environment defined
          if (environment === ENVIRONMENT_NOT_DEFINED.value) {
            return SERVICE_ENVIRONMENT in node;
          }

          // no match if environment is given but doesn't match the node's environment
          return node[SERVICE_ENVIRONMENT] !== environment;
        }

        return true;
      });
```

Reply:
I think we should also check for `node['service.environment'] === null`.

Reply:
Y'all are probably right. I approved it because it behaved correctly when I checked it out in this situation.